src/optimization/DAG.cpp

```diff
@@ -0,0 +1,149 @@
+//
+// Created by nomaddo on 17/12/18.
+//
+
+#include <log.h>
+#include <assert.h>
+#include <iostream>
+#include <functional>
+#include <algorithm>
+#include <string>
+#include <stack>
+#include <utility>
+#include <vector>
+
+#include "DAG.h"
+#include "../InstructionWalker.h"
+#include "log.h"
+
+#include "../ControlFlowGraph.h"
+#include "../DebugGraph.h"
+#include "../InstructionWalker.h"
+#include "../Profiler.h"
+
+namespace vc4c {
+
+/* XXX should be replaced with common routines that generate such instructions? */
+/* reading from mutex I/O register means getting mutex */
+static bool isGetLocInstruction (IL * instr) {
+	for (Value v: instr->getArguments()) {
+		if (v.hasRegister(REG_MUTEX)) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/* writing to mutex I/O register means release of mutex */
+static bool isReleaseLocInstruction (IL * instr) {
+	if (instr->getOutput().has_value()) {
+		auto output = instr->getOutput().value();
+		return output.hasRegister(REG_MUTEX);
+	}
+
+	return false;
+}
+
+DAG::DAG(BasicBlock &bb) {
+	std::map<Value, IL*> map;
```

Comment:
Just a note: a `Value` is far from unique, you will most likely get a lot of collisions here (e.g. all instructions writing into the same register/local).

If this is on purpose (e.g. only store the last writing of a local, since the following reads only depend on this one write), please add a comment to the fact (here or in line 139).